README.md
# group-me

This is a fork of the [Premade Groups Filter](https://mods.curse.com/addons/wow/premade-groups-filter) World of Warcraft addon, made by Elotheon on Arthas-EU. Most of the credit for this code goes to them.

I have added a few convenience features:
* Automatically searches for groups
* Automatically applies for groups using a chosen role


CHANGES.txt
lightweight tag 4ea6579b832ab87c6ce2b157602e84d88d8cf68b 1.6
Author:	Bernhard Saumweber <b.saumweber@gmail.com>
Date:	Sun Jan 1 21:19:33 2017 +0100

commit 4ea6579b832ab87c6ce2b157602e84d88d8cf68b
Author: Bernhard Saumweber <b.saumweber@gmail.com>
Date:   Sun Jan 1 21:19:33 2017 +0100

    Set version to 1.6 (keep forgetting that)

commit fc807b0db7553da94525b0a6124c216d33e220ad
Author: Bernhard Saumweber <b.saumweber@gmail.com>
Date:   Sun Jan 1 21:13:07 2017 +0100

    Added a tooltip to the new enable/disable Premade Groups Filter checkbox.

commit 503425ca00f5b9ec0480a78579207a9e6e46dee0
Author: Bernhard Saumweber <b.saumweber@gmail.com>
Date:   Sun Jan 1 20:34:02 2017 +0100

    Improved difficulty detection for legacy dungeons and raids. Added keywords activityname (string), maxplayers (integer), suggestedilvl (integer), minlvl (integer),  voicechat (string).

commit 0e339c156e3966ceeac61f601fe88dd531a205b3
Author: Bernhard Saumweber <b.saumweber@gmail.com>
Date:   Sun Jan 1 16:46:16 2017 +0100

    The new PGF checkbox can now be used to temporarily disable filtering without loosing the current filter settings. Fixed an addon loading issue. Moved some functions from Dialog\Dialog.lua to Dialog\Setup.lua and vice versa.



Init.lua
-------------------------------------------------------------------------------
-- Premade Groups Filter
-------------------------------------------------------------------------------
-- Copyright (C) 2015 Elotheon-Arthas-EU
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
-------------------------------------------------------------------------------

PremadeGroupsFilter = {}
PremadeGroupsFilterState = PremadeGroupsFilterState or {}

local PGFAddonName = select(1, ...)
local PGF = select(2, ...)

PremadeGroupsFilter.Debug = PGF

PGF.L = {}
PGF.C = {}

local L = PGF.L
local C = PGF.C

C.SAVED_STATE_VERSION = 14

C.NORMAL     = 1
C.HEROIC     = 2
C.MYTHIC     = 3
C.MYTHICPLUS = 4

-- corresponds to the third parameter of C_LFGList.GetActivityInfo()
C.TYPE_QUESTING = 1
C.TYPE_DUNGEON  = 2
C.TYPE_RAID     = 3
C.TYPE_ARENA    = 4
C.TYPE_SCENARIO = 5
C.TYPE_CUSTOM   = 6
C.TYPE_SKIRMISH = 7
C.TYPE_BG       = 8
C.TYPE_RBG      = 9
C.TYPE_ASHRAN   = 10

C.DIFFICULTY_STRING = {
    [1] = "normal",
    [2] = "heroic",
    [3] = "mythic",
    [4] = "mythicplus",
}

C.COLOR_ENTRY_NEW       = { R = 0.3, G = 1.0, B = 0.3 }
C.COLOR_LOCKOUT_PARTIAL = { R = 1.0, G = 0.5, B = 0.1 }
C.COLOR_LOCKOUT_FULL    = { R = 1.0, G = 0.1, B = 0.1 }

C.FONTSIZE_TEXTBOX = 12
C.SEARCH_ENTRY_RESET_WAIT = 2 -- wait at least 2 seconds between two resets of known premade groups

C.ROLE_PREFIX = {
    ["DAMAGER"] = "dps",
    ["HEALER"] = "heal",
    ["TANK"] = "tank",
}

C.ROLE_SUFFIX = {
    ["DAMAGER"] = "dps",
    ["HEALER"] = "heals",
    ["TANK"] = "tanks",
}

C.MODEL_DEFAULT = {
    enabled = true,
    expression = "",
    difficulty = {
        act = false,
        val = 3,
    },
    ilvl = {
        act = false,
        min = "",
        max = "",
    },
    noilvl = {
        act = false
    },
    members = {
        act = false,
        min = "",
        max = "",
    },
    tanks = {
        act = false,
        min = "",
        max = "",
    },
    heals = {
        act = false,
        min = "",
        max = "",
    },
    dps = {
        act = false,
        min = "",
        max = "",
    },
    defeated = {
        act = false,
        min = "",
        max = "",
    },
    tankrole = {
        act = false
    },
    healrole = {
        act = false
    },
    dpsrole = {
        act = false
    },
}

function PGF.OnAddonLoaded(name)
    if name == PGFAddonName then
        PGF.Table_UpdateWithDefaults(PremadeGroupsFilterState, PGF.C.MODEL_DEFAULT)
    end
end

function PGF.OnEvent(self, event, ...)
    if event == "ADDON_LOADED" then PGF.OnAddonLoaded(...) end
end

local frame = CreateFrame("Frame", "PremadeGroupsFilterEventFrame")
frame:RegisterEvent("ADDON_LOADED")
frame:SetScript("OnEvent", PGF.OnEvent)


Dialog/Setup.lua
-------------------------------------------------------------------------------
-- Premade Groups Filter
-------------------------------------------------------------------------------
-- Copyright (C) 2015 Elotheon-Arthas-EU
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
-------------------------------------------------------------------------------

local PGF = select(2, ...)
local L = PGF.L
local C = PGF.C

local DIFFICULTY_TEXT = {
    [C.NORMAL] = L["dialog.normal"],
    [C.HEROIC] = L["dialog.heroic"],
    [C.MYTHIC] = L["dialog.mythic"],
    [C.MYTHICPLUS] = L["dialog.mythicplus"],
}

-------------------------------------------------------------------------------
-- OnShow functions
-------------------------------------------------------------------------------

function PGF.Dialog_LoadMinMaxFromModel(dialog, model, key)
    dialog[key].Act:SetChecked(model[key:lower()].act)
    dialog[key].Min:SetText(model[key:lower()].min)
    dialog[key].Max:SetText(model[key:lower()].max)
end

function PGF.Dialog_LoadFromModel(dialog)
    local model = PremadeGroupsFilterState
    PGF.UsePFGButton:SetChecked(model.enabled)
    PGF.previousSearchExpression = model.expression
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Ilvl")
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Defeated")
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Members")
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Tanks")
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Heals")
    PGF.Dialog_LoadMinMaxFromModel(dialog, model, "Dps")
    dialog.Noilvl.Act:SetChecked(model.noilvl.act)
    dialog.Tankrole.Act:SetChecked(model.tankrole.act)
    dialog.Healrole.Act:SetChecked(model.healrole.act)
    dialog.Dpsrole.Act:SetChecked(model.dpsrole.act)
    dialog.Expression.EditBox:SetText(model.expression)
    dialog.Difficulty.Act:SetChecked(model.difficulty.act)
    UIDropDownMenu_SetSelectedValue(dialog.Difficulty.DropDown, model.difficulty.val)
    UIDropDownMenu_SetText(dialog.Difficulty.DropDown, DIFFICULTY_TEXT[model.difficulty.val])
end

function PGF.Dialog_UpdatePosition()
    local dialog = PremadeGroupsFilterDialog
    dialog:SetPoint("TOPLEFT", GroupFinderFrame, "TOPRIGHT")
    dialog:SetPoint("BOTTOMLEFT", GroupFinderFrame, "BOTTOMRIGHT")
    dialog:SetWidth(300)
end

function PGF.Dialog_OnShow(dialog)
    RequestRaidInfo() -- need the dungeon/raid lockout information later for filtering
    PGF.Dialog_LoadFromModel(dialog)
    PGF.Dialog_UpdatePosition(dialog)
end

-------------------------------------------------------------------------------
-- OnLoad functions
-------------------------------------------------------------------------------

function PGF.Dialog_DifficultyDropdown_AddItem(dropdown, difficulty, text)
    local info = UIDropDownMenu_CreateInfo()
    info.value = difficulty
    info.checked = false -- we do not have settings at this point and will set checked value later anyway
    info.func = PGF.Dialog_DifficultyDropdown_OnClick
    info.arg1 = dropdown
    info.arg2 = text
    info.text = text
    UIDropDownMenu_AddButton(info)
end

function PGF.Dialog_DifficultyDropdown_Init(self, level)
    PGF.Dialog_DifficultyDropdown_AddItem(self, C.NORMAL, DIFFICULTY_TEXT[C.NORMAL])
    PGF.Dialog_DifficultyDropdown_AddItem(self, C.HEROIC, DIFFICULTY_TEXT[C.HEROIC])
    PGF.Dialog_DifficultyDropdown_AddItem(self, C.MYTHIC, DIFFICULTY_TEXT[C.MYTHIC])
    PGF.Dialog_DifficultyDropdown_AddItem(self, C.MYTHICPLUS, DIFFICULTY_TEXT[C.MYTHICPLUS])
end

function PGF.Dialog_SetUpGenericField(self, key)
    self[key]:SetAttribute("parentKey", key)
    self[key].Title:SetText(L["dialog." .. key:lower()])
    self[key].Act:SetScript("OnClick", PGF.Dialog_Act_OnClick)
end

function PGF.Dialog_SetUpMinMaxField(self, key)
    PGF.Dialog_SetUpGenericField(self, key)
    self[key].Min:SetAttribute("parentKey", "Min")
    self[key].Max:SetAttribute("parentKey", "Max")
    self[key].To:SetText(L["dialog.to"])
    self[key].Min:SetScript("OnTextChanged", PGF.Dialog_MinMax_OnTextChanged)
    self[key].Max:SetScript("OnTextChanged", PGF.Dialog_MinMax_OnTextChanged)
    self[key].Min:SetScript("OnTabPressed", PGF.Dialog_Min_OnTabPressed)
    self[key].Max:SetScript("OnTabPressed", PGF.Dialog_Max_OnTabPressed)
end

function PGF.Dialog_SetUpUsePGFCheckbox()
    local button = CreateFrame("CheckButton", "UsePFGButton", LFGListFrame.SearchPanel, "UICheckButtonTemplate")
    button:SetSize(26, 26)
    button:SetHitRectInsets(-2, -30, -2, -2)
    button.text:SetText("PGF")
    button.text:SetFontObject("GameFontHighlight")
    button.text:SetWidth(30)
    button:SetPoint("LEFT", LFGListFrame.SearchPanel.FilterButton, "LEFT", 0, 0)
    button:SetPoint("TOP", LFGListFrame.SearchPanel.RefreshButton, "TOP", 0, -3)
    button:SetScript("OnClick", PGF.Dialog_UsePGF_OnClick)
    button:SetScript("OnEnter", function (self)
        GameTooltip:SetOwner(self, "ANCHOR_RIGHT")
        GameTooltip:SetText(L["dialog.usepgf.tooltip"])
    end)
    button:SetScript("OnLeave", function () GameTooltip:Hide() end)
    PGF.UsePFGButton = button
end

function PGF.Dialog_OnLoad()
    local dialog = PremadeGroupsFilterDialog -- keep that
    dialog:SetScript("OnShow", PGF.Dialog_OnShow)
    dialog:SetScript("OnUpdate", PGF.Dialog_OnUpdate)
    dialog.TimeSinceLastUpdate = 0
    dialog.searching = false

    dialog.InsetBg:SetPoint("TOPLEFT", 4, -62)
    dialog.InsetBg:SetPoint("BOTTOMRIGHT", -6, 26)
    dialog.Title:SetText("Premade Groups Filter")
    dialog.ResetButton:SetText(L["dialog.reset"])
    dialog.ResetButton:SetScript("OnClick", PGF.Dialog_ResetButton_OnClick)
    dialog.RefreshButton:SetText(L["dialog.refresh"])
    dialog.RefreshButton:SetScript("OnClick", PGF.Dialog_RefreshButton_OnClick)
    dialog.SimpleExplanation:SetText(L["dialog.expl.simple"])
    dialog.StateExplanation:SetText(L["dialog.expl.state"])
    dialog.MinExplanation:SetText(L["dialog.expl.min"])
    dialog.MaxExplanation:SetText(L["dialog.expl.max"])
    dialog.Advanced.Explanation:SetText(L["dialog.expl.advanced"])
    dialog.Advanced.InfoButton:EnableMouse(true)
    dialog.Advanced.InfoButton:SetScript("OnEnter", PGF.Dialog_InfoButton_OnEnter)
    dialog.Advanced.InfoButton:SetScript("OnLeave", PGF.Dialog_InfoButton_OnLeave)
    dialog.Ilvl.Min:SetMaxLetters(3)
    dialog.Ilvl.Max:SetMaxLetters(3)
    dialog.Noilvl.Title:SetWidth(210)
    dialog.Noilvl.Act:SetEnabled(false)
    dialog.Defeated.Title:SetWordWrap(true)
    dialog.Defeated.Title:SetHeight(28)

    PGF.Dialog_SetUpGenericField(dialog, "Difficulty")
    PGF.Dialog_SetUpMinMaxField(dialog, "Ilvl")
    PGF.Dialog_SetUpGenericField(dialog, "Noilvl")
    PGF.Dialog_SetUpMinMaxField(dialog, "Members")
    PGF.Dialog_SetUpMinMaxField(dialog, "Tanks")
    PGF.Dialog_SetUpMinMaxField(dialog, "Heals")
    PGF.Dialog_SetUpMinMaxField(dialog, "Dps")
    PGF.Dialog_SetUpMinMaxField(dialog, "Defeated")
    PGF.Dialog_SetUpGenericField(dialog, "Tankrole")
    PGF.Dialog_SetUpGenericField(dialog, "Healrole")
    PGF.Dialog_SetUpGenericField(dialog, "Dpsrole")
    PGF.Dialog_SetUpUsePGFCheckbox()

    local font = dialog.SimpleExplanation:GetFont()
    dialog.Expression.EditBox:SetFont(font, C.FONTSIZE_TEXTBOX)
    dialog.Expression.EditBox.Instructions:SetFont(font, C.FONTSIZE_TEXTBOX)
    --dialog.Expression.EditBox:SetScript("OnTextChanged", PGF.Dialog_Expression_OnTextChanged) -- overrides Blizz

    UIDropDownMenu_Initialize(dialog.Difficulty.DropDown, PGF.Dialog_DifficultyDropdown_Init)
    UIDropDownMenu_SetWidth(dialog.Difficulty.DropDown, 90)
end

function PGF.Dialog_OnUpdate(this, arg1)
  PremadeGroupsFilterDialog.TimeSinceLastUpdate = PremadeGroupsFilterDialog.TimeSinceLastUpdate + arg1
  if (PremadeGroupsFilterDialog.TimeSinceLastUpdate > 10 and PremadeGroupsFilterDialog.searching) then
    LFGListSearchPanel_DoSearch(LFGListFrame.SearchPanel)
		PremadeGroupsFilterDialog.TimeSinceLastUpdate = 0
	end
end

PremadeGroupsFilter.Dialog_OnLoad = PGF.Dialog_OnLoad


Main.lua
-------------------------------------------------------------------------------
-- Premade Groups Filter
-------------------------------------------------------------------------------
-- Copyright (C) 2015 Elotheon-Arthas-EU
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
-------------------------------------------------------------------------------

local PGF = select(2, ...)
local L = PGF.L
local C = PGF.C

PGF.lastSearchEntryReset = time()
PGF.previousSearchExpression = ""
PGF.currentSearchExpression = ""
PGF.previousSearchLeaders = {}
PGF.currentSearchLeaders = {}
PGF.signedUpGroups = {}

function PGF.GetExpressionFromMinMaxModel(model, key)
    local exp = ""
    if model[key].act then
        if PGF.NotEmpty(model[key].min) then exp = exp .. " and " .. key .. ">=" .. model[key].min end
        if PGF.NotEmpty(model[key].max) then exp = exp .. " and " .. key .. "<=" .. model[key].max end
    end
    return exp
end

function PGF.GetExpressionFromIlvlModel(model)
    local exp = PGF.GetExpressionFromMinMaxModel(model, "ilvl")
    if model.noilvl.act and PGF.NotEmpty(exp) then
        exp = exp:gsub("^ and ", "")
        exp = " and (" .. exp .. " or ilvl==0)"
    end
    return exp
end

function PGF.GetExpressionFromDifficultyModel(model)
    if model.difficulty.act then
        return " and " .. C.DIFFICULTY_STRING[model.difficulty.val]
    end
    return ""
end

function PGF.GetExpressionFromAdvancedExpression(model)
    if model.expression and model.expression ~= "" then
        return " and " .. model.expression
    end
    return ""
end

function PGF.GetExpressionFromModel()
    local model = PremadeGroupsFilterState
    local exp = "true" -- start with neutral element
    exp = exp .. PGF.GetExpressionFromDifficultyModel(model)
    exp = exp .. PGF.GetExpressionFromIlvlModel(model)
    exp = exp .. PGF.GetExpressionFromMinMaxModel(model, "members")
    exp = exp .. PGF.GetExpressionFromMinMaxModel(model, "tanks")
    exp = exp .. PGF.GetExpressionFromMinMaxModel(model, "heals")
    exp = exp .. PGF.GetExpressionFromMinMaxModel(model, "dps")
    exp = exp .. PGF.GetExpressionFromMinMaxModel(model, "defeated")
    exp = exp .. PGF.GetExpressionFromAdvancedExpression(model)
    exp = exp:gsub("^true and ", "")
    return exp
end

function PGF.ResetSearchEntries()
    -- make sure to wait at least some time between two resets
    if time() - PGF.lastSearchEntryReset > C.SEARCH_ENTRY_RESET_WAIT then
        PGF.previousSearchLeaders = PGF.Table_Copy_Shallow(PGF.currentSearchLeaders)
        PGF.currentSearchLeaders = {}
        PGF.previousSearchExpression = PGF.currentSearchExpression
        PGF.lastSearchEntryReset = time()
    end
end

function PGF.SortByFriendsAndAge(id1, id2)
    local _, _, _, _, _, _, _, age1, bnetFriends1, charFriends1, guildMates1 = C_LFGList.GetSearchResultInfo(id1);
    local _, _, _, _, _, _, _, age2, bnetFriends2, charFriends2, guildMates2 = C_LFGList.GetSearchResultInfo(id2);
    if bnetFriends1 ~= bnetFriends2 then return bnetFriends1 > bnetFriends2 end
    if charFriends1 ~= charFriends2 then return charFriends1 > charFriends2 end
    if guildMates1 ~= guildMates2 then return guildMates1 > guildMates2 end
    return age1 < age2
end

function PGF.OnLFGListSortSearchResults(results)
    PGF.ResetSearchEntries()
    local exp = PGF.GetExpressionFromModel()
    PGF.currentSearchExpression = exp
    if not PremadeGroupsFilterState.enabled then return end
    if exp == "true" then return end -- skip trivial expression

    -- loop backwards through the results list so we can remove elements from the table
    for idx = #results, 1, -1 do
        local resultID = results[idx]
        local _, activity, name, comment, voiceChat, iLvl, honorLevel, age,
              numBNetFriends, numCharFriends, numGuildMates, _, leaderName,
              numMembers = C_LFGList.GetSearchResultInfo(resultID)
        local completedEncounters = C_LFGList.GetSearchResultEncounterInfo(resultID)
        local memberCounts = C_LFGList.GetSearchResultMemberCounts(resultID)
        local partialLockout, fullLockout = PGF.HasDungeonOrRaidLockout(activity)
        local avName, avShortName, avCategoryID, avGroupID, avILevel, avFilters,
              avMinLevel, avMaxPlayers, avDisplayType, avOrderIndex,
              avUseHonorLevel, avShowQuickJoin = C_LFGList.GetActivityInfo(activity)
        local difficulty = PGF.GetDifficulty(activity, avName, avShortName)

        local env = {}
        env.activity = activity
        env.name = name:lower()
        env.activityname = avName:lower()
        env.comment = comment:lower()
        env.leader = leaderName and leaderName:lower() or ""
        env.age = math.floor(age / 60) -- age in minutes
        env.voice = voiceChat and voiceChat ~= ""
        env.voicechat = voiceChat
        env.ilvl = iLvl or 0
        env.hlvl = honorLevel or 0
        env.friends = numBNetFriends + numCharFriends + numGuildMates
        env.members = numMembers
        env.tanks = memberCounts.TANK
        env.heals = memberCounts.HEALER
        env.dps = memberCounts.DAMAGER + memberCounts.NOROLE
        env.defeated = completedEncounters and #completedEncounters or 0
        env.normal     = difficulty == C.NORMAL
        env.heroic     = difficulty == C.HEROIC
        env.mythic     = difficulty == C.MYTHIC
        env.mythicplus = difficulty == C.MYTHICPLUS
        env.myrealm = leaderName and not leaderName:find('-')
        env.partialid = partialLockout
        env.fullid = fullLockout
        env.noid = not fullLockout and not partialLockout
        env.maxplayers = avMaxPlayers
        env.suggestedilvl = avILevel
        env.minlvl = avMinLevel
        env.categoryid = avCategoryID
        env.groupid = avGroupID

        for i = 1, numMembers do
            local role, class = C_LFGList.GetSearchResultMemberInfo(resultID, i);
            local classPlural = class:lower() .. "s" -- plural form of the class in english
            env[classPlural] = (env[classPlural] or 0) + 1
            if role then
                local classRolePlural = C.ROLE_PREFIX[role] .. "_" .. class:lower() .. "s"
                local roleClassPlural = class:lower() .. "_" .. C.ROLE_SUFFIX[role]
                env[classRolePlural] = (env[classRolePlural] or 0) + 1
                env[roleClassPlural] = (env[roleClassPlural] or 0) + 1
            end
        end

        env.arena2v2 = activity == 6
        env.arena3v3 = activity == 7

        -- raids            normal             heroic             mythic
        env.hm   = activity ==  37 or activity ==  38 or activity == 399  -- Highmaul
        env.brf  = activity ==  39 or activity ==  40 or activity == 400  -- Blackrock Foundry
        env.hfc  = activity == 409 or activity == 410 or activity == 412  -- Hellfire Citadel
        env.en   = activity == 413 or activity == 414                     -- The Emerald Nightmare
        env.nh   = activity == 415 or activity == 416                     -- The Nighthold
        env.tov  = activity == 456 or activity == 457                     -- Trial of Valor

        -- dungeons         normal             heroic             mythic            mythic+
        env.eoa  = activity == 425 or activity == 435 or activity == 445 or activity == 459  -- Eye of Azshara
        env.dht  = activity == 426 or activity == 436 or activity == 446 or activity == 460  -- Darkheart Thicket
        env.hov  = activity == 427 or activity == 437 or activity == 447 or activity == 461  -- Halls of Valor
        env.nl   = activity == 428 or activity == 438 or activity == 448 or activity == 462  -- Neltharion's Lair
        env.vh   = activity == 429 or activity == 439 or activity == 449                     -- Violet Hold
        env.brh  = activity == 430 or activity == 440 or activity == 450 or activity == 463  -- Black Rook Hold
        env.votw = activity == 431 or activity == 441 or activity == 451 or activity == 464  -- Vault of the Wardens
        env.mos  = activity == 432 or activity == 442 or activity == 452 or activity == 465  -- Maw of Souls
        env.cos  = activity == 433 or activity == 443 or activity == 453 or activity == 466  -- Court of Stars
        env.aw   = activity == 434 or activity == 444 or activity == 454 or activity == 467  -- The Arcway
        env.kara =                                       activity == 455                     -- Karazhan

        setmetatable(env, { __index = function(table, key) return 0 end }) -- set non-initialized values to 0
        if PGF.DoesPassThroughFilter(env, exp) then
            -- leaderName is usually still nil at this point if the group is new, but we can live with that
            if leaderName then PGF.currentSearchLeaders[leaderName] = true end
        else
            table.remove(results, idx)
        end
    end
    -- sort by age
    table.sort(results, PGF.SortByFriendsAndAge)
    LFGListFrame.SearchPanel.totalResults = #results
end

function PGF.OnLFGListSearchEntryUpdate(self)
    local id, activity, _, _, _, _, _, _, _, _, _, isDelisted, leaderName = C_LFGList.GetSearchResultInfo(self.resultID)
    -- try once again to update the leaderName (this information is not immediately available)
    if leaderName then PGF.currentSearchLeaders[leaderName] = true end
    if not isDelisted then
        -- color name if new
        if PGF.currentSearchExpression ~= "true"                        -- not trivial search
        and PGF.currentSearchExpression == PGF.previousSearchExpression -- and the same search
        and (leaderName and not PGF.previousSearchLeaders[leaderName]) then              -- and leader is new
            local color = C.COLOR_ENTRY_NEW
            self.Name:SetTextColor(color.R, color.G, color.B);
            if PGF.signedUpGroups[id] == nil and PremadeGroupsFilterDialog.searching and C_LFGList.GetRoleCheckInfo() == false then
                print("signed up to ", id)
                PGF.signedUpGroups[id] = true
                C_LFGList.ApplyToGroup(id, "", PremadeGroupsFilterState.tankrole.act, PremadeGroupsFilterState.healrole.act, PremadeGroupsFilterState.dpsrole.act) -- SIGN UP AS CHOSEN ROLE
            end
        end
        -- color activity if lockout
        local partialLockout, fullLockout = PGF.HasDungeonOrRaidLockout(activity)
        if fullLockout then
            local color = C.COLOR_LOCKOUT_FULL
            self.ActivityName:SetTextColor(color.R, color.G, color.B);
        elseif partialLockout then
            local color = C.COLOR_LOCKOUT_PARTIAL
            self.ActivityName:SetTextColor(color.R, color.G, color.B);
        end
    end
end

function PGF.OnLFGListSearchEntryOnEnter(self)
    local resultID = self.resultID
    local _, _, _, _, _, _, _, _, _, _, _, isDelisted, _, numMembers = C_LFGList.GetSearchResultInfo(resultID)
    if isDelisted or not GameTooltip:IsShown() then return end
    GameTooltip:AddLine(" ")
    GameTooltip:AddLine(CLASS_ROLES);

    local roles = {}
    local classInfo = {}
    for i = 1, numMembers do
        local role, class, classLocalized = C_LFGList.GetSearchResultMemberInfo(resultID, i);
        classInfo[class] = {
            name = classLocalized,
            color = RAID_CLASS_COLORS[class] or NORMAL_FONT_COLOR
        }
        if not roles[role] then roles[role] = {} end
        if not roles[role][class] then roles[role][class] = 0 end
        roles[role][class] = roles[role][class] + 1
    end

    for role, classes in pairs(roles) do
        GameTooltip:AddLine(_G[role]..": ")
        for class, count in pairs(classes) do
            local text = "   "
            if count > 1 then text = text .. count .. " " else text = text .. "   " end
            text = text .. "|c" .. classInfo[class].color.colorStr ..  classInfo[class].name .. "|r "
            GameTooltip:AddLine(text)
        end
    end
    GameTooltip:Show()
end

function PGF.OnLFGListSearchEntryOnClick(self)
    PGF.Dialog_StopSearch()
end

hooksecurefunc("LFGListUtil_SortSearchResults", PGF.OnLFGListSortSearchResults)
hooksecurefunc("LFGListSearchEntry_Update", PGF.OnLFGListSearchEntryUpdate)
hooksecurefunc("LFGListSearchEntry_OnEnter", PGF.OnLFGListSearchEntryOnEnter)
hooksecurefunc("LFGListSearchEntry_OnClick", PGF.OnLFGListSearchEntryOnClick)
